dll/ntdll/ldr/appcompat.c

```diff
@@ -0,0 +1,386 @@
+{
+    PIMAGE_SECTION_HEADER SectionHeader;
+
+    SectionHeader = LdrpFindSectionByName(ImageBase, ".expvers");
+    if (SectionHeader == NULL)
+    {
+        return NULL;
+    }
+
+    return (PROSCOMPAT_DESCRIPTOR)((PUCHAR)ImageBase + SectionHeader->VirtualAddress);
+}
+
+static
+NTSTATUS
+GetExportNameVersionTable(
+    _In_ PVOID ImageBase,
+    _Out_ PULONG *Table,
+    _Out_ PULONG NumberOfEntries)
+{
+    PROSCOMPAT_DESCRIPTOR RosCompatDescriptor;
+
+    RosCompatDescriptor = FindRosCompatDescriptor(ImageBase);
+    if (RosCompatDescriptor == NULL)
+    {
+        return STATUS_NOT_FOUND;
+    }
+
+    *NumberOfEntries = RosCompatDescriptor->NumberOfExportNames;
+    *Table = RosCompatDescriptor->ExportNameMasks;
+
+    return STATUS_SUCCESS;
+}
+
+__inline
+BOOLEAN IsPowerOfTwo(unsigned int x)
+{
+    return ((x != 0) && !(x & (x - 1)));
+}
+
+static
+NTSTATUS
+LdrpPatchExportNameTable(
+    _In_ PVOID DllBase,
+    _In_ PROSCOMPAT_DESCRIPTOR RosCompatDescriptor,
+    _Inout_ PIMAGE_EXPORT_DIRECTORY ExportDirectory,
+    _In_ DWORD AppCompatVersion)
+{
+    ULONG VersionMask;
+    PULONG NameTable, OrgNameTable;
+    PUSHORT OrdinalTable, OrgOrdinalTable;
+    SIZE_T Size;
+    ULONG i, j, k;
+
+    /* Translate to version mask */
+    VersionMask = 1 << TranslateAppcompatVersionToVersionBit(AppCompatVersion);
+    ASSERT(IsPowerOfTwo(VersionMask));
+
+    /* Get the VA of the name table */
+    NameTable = (PULONG)((ULONG_PTR)DllBase +
+        (ULONG_PTR)ExportDirectory->AddressOfNames);
+
+    OrdinalTable = (PUSHORT)((ULONG_PTR)DllBase +
+        (ULONG_PTR)ExportDirectory->AddressOfNameOrdinals);
+
+    /* Allocate a temp buffer for the exports */
+    Size = ExportDirectory->NumberOfNames * (sizeof(ULONG) + sizeof(USHORT));
+    OrgNameTable = RtlAllocateHeap(RtlGetProcessHeap(), 0, Size);
+    if (OrgNameTable == NULL)
+    {
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    OrgOrdinalTable = (PUSHORT)&OrgNameTable[ExportDirectory->NumberOfNames];
+
+    /* Make a copy of the original exports */
+    RtlCopyMemory(OrgNameTable, NameTable, ExportDirectory->NumberOfNames * sizeof(ULONG));
+    RtlCopyMemory(OrgOrdinalTable, OrdinalTable, ExportDirectory->NumberOfNames * sizeof(USHORT));
+
+    /* Strip unused entries from the name and ordinal table */
+    for (i = 0, j = 0, k = 0; i < ExportDirectory->NumberOfNames; i++)
+    {
+        if (RosCompatDescriptor->ExportNameMasks[i] & VersionMask)
+        {
+            /* Put public functions into the export table */
+            NameTable[j] = OrgNameTable[i];
+            OrdinalTable[j] = OrgOrdinalTable[i];
+            j++;
+        }
+        else
+        {
+            /* Move private functions to the start of the temp buffer */
+            OrgNameTable[k] = OrgNameTable[i];
+            OrgOrdinalTable[k] = OrgOrdinalTable[i];
+            k++;
+        }
+    }
+
+    /* Finally patch the number of exports */
+    ExportDirectory->NumberOfNames = j;
```

Comment:
We probably also need to patch `NumberOfFunctions` here, but that would also mean adding some code to move the 'hidden' exports behind ordinal-only exports,
and thus re-numbering our exports?

Reply:
I don't think so. There can always be more functions than names (nameless exports). Import by ordinal will not be changed for now. I don't think that this will be a problem. If yes, then the name table needs to be patched as well.